aten/src/ATen/native/cuda/Dropout.cu

```diff
@@ -149,7 +149,7 @@ fused_dropout_kernel(cuda::detail::TensorInfo<scalar_t, IndexType> a,
            if (li < totalElements) {
     // Convert `linearIndex` into an offset of `b`
                const IndexType bOffset =
-                   cuda::detail::IndexToOffset<scalar_t, IndexType, 1>::get(li, b);
+                   cuda::detail::IndexToOffset<scalar_t, IndexType, ADims>::get(li, b);
```

Comment:
This is not a correct fix. In most cases mask and output will be contiguous for discontiguous input, and the dim here needs to be 1 (they are coalesced to 1). Only in the case where `suggest_memory_format` is channels_last and for some reason vectorized kernel is not used will this error happen.

Reply:
https://github.com/pytorch/pytorch/blob/5a5258cb0df5f57640c5af0e2d36ef5c5f97cfb6/aten/src/ATen/native/cuda/Dropout.cu#L225-L230

Post this I just printed the dims for each Tensor
```c++
      std::cout << self_info.dims << " " << mask_info.dims << " " << ret_info.dims << "\n";
      auto print_dims = [](auto tensor_info) {
        for (int i = 0; i < tensor_info.dims; ++i) {
          std::cout << tensor_info.sizes[i] << " ";
        }
        std::cout << "\n";
      };
      std::cout << "SELF_INFO SIZES:";
      print_dims(self_info);
      std::cout << "MASK_INFO SIZES:";
      print_dims(mask_info);
      std::cout << "RET_INFO SIZES:";
      print_dims(ret_info);
```

And the output is,
```
3 3 3
SELF_INFO SIZES:2 3 9 
MASK_INFO SIZES:2 3 9 
RET_INFO SIZES:2 3 9 
```

This is why I believe, replacing `1` with `ADims` is correct.

Reply:
An example in the issue is one of possible failure modes. In other cases input could be discontiguous (not collapsible), but mask/output will be collapsible. In this case ADim will be -1, so using -1 for output will work too, but it's less efficient than 1.

Reply:
https://github.com/pytorch/pytorch/blob/5a5258cb0df5f57640c5af0e2d36ef5c5f97cfb6/aten/src/ATen/native/cuda/Dropout.cu#L200-L203

Also printing the strides after mask using (for the case in the issue with self (2, 3, 3, 3))
```c++
  std::cout << "SELF STRIDES:" << self.strides() << "\n";
  std::cout << "RET STRIDES:" << ret.strides() << "\n";
  std::cout << "MASK STRIDES:" << mask.strides() << "\n";
  std::cout << "RET CONTIGUOUS:" << ret.is_contiguous() << "\n";
  std::cout << "MASK CONTIGUOUS:" << mask.is_contiguous() << "\n";
```

Output
```
SELF STRIDES:[54, 1, 18, 6]
RET STRIDES:[27, 1, 9, 3]
MASK STRIDES:[27, 1, 9, 3]
RET CONTIGUOUS:0
MASK CONTIGUOUS:0
```
Note `ret` and `mask` are also not contiguous and their stride are similar to `self`

Reply:
> An example in the issue is one of possible failure modes. In other cases input could be discontiguous (not collapsible), but mask/output will be collapsible. In this case ADim will be -1, so using -1 for output will work too, but it's less efficient than 1.

I think `mask` and `ret` would mostly preserve the behavior that input has :thinking:

Reply:
They don't 
```
import torch
mod = torch.nn.Dropout(0.5)
inp = torch.randn(3,4, device="cuda").t()
out=mod(inp)
print(inp.stride(), out.stride()) # (1, 4) (3, 1), output is contiguous, input is not
inp = torch.randn(3,4, device="cuda")[:,::2]
out = mod(inp)
print(inp.stride(), out.stride()) # (4, 2) (2, 1) output is contiguous, input is not
```

Reply:
Ah. I see. Thank  you very much!

Reply:
Note:

After changing,
```c++
Tensor ret = at::empty_like(self, self.suggest_memory_format()); 
Tensor mask = at::empty(self.sizes(), self.options().dtype(kByte), self.suggest_memory_format()); 
```

to 
```C++
Tensor ret = at::empty_like(self);
Tensor mask = at::empty_like(self, self.options().dtype(kByte));
```

Output of above snippet is 
```
(1, 4) (1, 4)
(4, 2) (2, 1)
```

Reply:
Yep, that's what we want to happen - dropout preserves strides of input if input is memory-dense, and makes output dense while preserving input permutation if input is not memory-dense. 
Note also that the first case can go to vectorized kernel now.